File: admin/src/utils/fileHash.ts
/**
 * 文件哈希工具（Admin）
 *
 * 使用浏览器内置 Web Crypto 计算 SHA-256，
 * 对应后端的 hash_alg = 2。
 * 若运行环境不支持 Web Crypto，则优雅降级为不带哈希。
 */

export interface FileHashResult {
  hashValue: string | null;
  hashAlg: number | null;
}

/**
 * 计算文件哈希（SHA-256）
 */
export async function computeFileHash(file: Blob): Promise<FileHashResult> {
  try {
    const cryptoObj: Crypto | undefined =
      (window as any).crypto || (window as any).msCrypto;
    if (
      !cryptoObj ||
      !cryptoObj.subtle ||
      typeof cryptoObj.subtle.digest !== 'function'
    ) {
      // 运行环境不支持 Web Crypto，直接降级
      // eslint-disable-next-line no-console
      console.warn('[admin:fileHash] crypto.subtle 不可用，跳过哈希计算');
      return { hashValue: null, hashAlg: null };
    }

    const buffer = await file.arrayBuffer();
    const digest = await cryptoObj.subtle.digest('SHA-256', buffer);
    const hashArray = Array.from(new Uint8Array(digest));
    const hashHex = hashArray
      .map((b) => b.toString(16).padStart(2, '0'))
      .join('');

    return {
      hashValue: hashHex,
      hashAlg: 2, // 2 = sha256
    };
  } catch (error) {
    // eslint-disable-next-line no-console
    console.warn('[admin:fileHash] 计算文件哈希失败，将跳过哈希上报', error);
    return { hashValue: null, hashAlg: null };
  }
}

File: app/lib/core/utils/file_hash.dart
import 'dart:typed_data';
import 'package:crypto/crypto.dart' as crypto;

/// 文件哈希结果
class FileHashResult {
  const FileHashResult({
    required this.hashValue,
    required this.hashAlg,
  });

  /// 十六进制哈希值（失败时为 null）
  final String? hashValue;

  /// 哈希算法编号：2 = sha256；失败或不支持时为 null
  final int? hashAlg;
}

/// 计算文件/二进制数据的 SHA-256 哈希（前端 Flutter 侧）  
/// 若运行环境不支持 crypto（如 web 未开启），调用方可选择忽略结果。
Future<FileHashResult> computeFileHash(Uint8List bytes) async {
  try {
    final digest = crypto.sha256.convert(bytes);
    final hashHex = digest.toString();
    return FileHashResult(hashValue: hashHex, hashAlg: 2);
  } catch (_) {
    // 兜底：无法计算时返回空结果
    return const FileHashResult(hashValue: null, hashAlg: null);
  }
}
